auraed/src/ebpf/tracepoint_programs/tracepoint_program.rs

```diff
@@ -0,0 +1,130 @@
+        // per-CPU buffers
+        let channel_capacity = per_cpu_buffer_capacity * num_cpus;
+
+        // Attach to kernel trace event
+        let _ = program.attach(Self::CATEGORY, Self::EVENT)?;
+
+        // Create the channel for broadcasting the events
+        let (tx, _) = broadcast::channel(channel_capacity);
+
+        // Open the BPF_PERF_EVENT_ARRAY BPF map that is used to send data from
+        // kernel to userspace. This array contains the per-CPU buffers and is
+        // indexed by CPU id.
+        // https://libbpf.readthedocs.io/en/latest/api.html
+        let mut perf_array =
+            AsyncPerfEventArray::try_from(bpf.map_mut(Self::PERF_BUFFER)?)?;
+
+        // Spawn a thread per CPU to listen for events from the kernel.
+        for cpu_id in online_cpus()? {
+            trace!("spawning task for cpu {cpu_id}");
+            // Open the per-CPU buffer for the current CPU id
+            let mut per_cpu_buffer =
+                perf_array.open(cpu_id, Some(PER_CPU_BUFFER_SIZE_IN_PAGES))?;
+
+            // Clone the sender of the event broadcast channel
+            let per_cpu_tx = tx.clone();
+
+            // Spawn the thread to listen on the per-CPU buffer
+            let _ignored = tokio::spawn(async move {
+                trace!("task for cpu {cpu_id} awaiting for events");
+
+                // Allocate enough memory to drain the entire buffer
+                // Note: using `vec!` macro will not result in a correct `Vec`
+                let mut buffers = (0..per_cpu_buffer_capacity)
+                    .map(|_| BytesMut::with_capacity(event_struct_size))
+                    .collect::<Vec<_>>();
+
+                // Start polling the per-CPU buffer for events
+                loop {
+                    let events = match per_cpu_buffer
+                        .read_events(&mut buffers)
+                        .await
+                    {
+                        Ok(events) => events,
+                        Err(error) => {
+                            error!("fail to read events from per-cpu perf buffer, bailing out: {error}");
+                            return;
+                        }
+                    };
+
+                    if events.lost > 0 {
+                        error!(
+                            "buffer full, dropped {} perf events - this should never happen!",
+                            events.lost
+                        );
+                    }
+
+                    for buf in buffers.iter_mut().take(events.read) {
+                        let ptr = buf.as_ptr() as *const T;
+                        let signal = unsafe { ptr.read_unaligned() };
+                        if per_cpu_tx.send(signal).is_err() {
+                            // send only errors if there are no listeners,
+                            // it can be safely ignored; future sends may succeed
+                        }
+                        buf.clear();
```

Comment:
I don't think there was an issue with the loop re-reading signals, but I thought it would be a good idea to clear the signals after they are processed

Reply:
As long as it's fast :) Again for POSIX signals this won't be a hot path, but for other future instrumentations, it will be.

Reply:
I'll take it out an leave a comment